ompi/mca/mtl/ofi/mtl_ofi.h

```diff
@@ -79,13 +80,14 @@ ompi_mtl_ofi_progress(void)
                 assert(ofi_req);
                 ret = ofi_req->event_callback(&wc, ofi_req);
                 if (OMPI_SUCCESS != ret) {
-                    opal_output(ompi_mtl_base_framework.framework_output,
-                                "Error returned by request event callback: %zd",
-                                ret);
-                    abort();
+                    fprintf(stderr, "Error returned by request event callback: %zd.\n"
+                                    "*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).\n",
+                                    ret);
+                    fflush(stderr);
```

Comment:
I would suggest `opal_show_help()` for all of these that you used `fprintf()` for, for 2 reasons:

1. `opal_output()` and `opal_show_help()` use a dedicated IO stream to send back messages to `mpirun`'s stderr.  Hence, they provide a level of atomicity that you don't get with normal `fprintf()` (because `fprintf()` contents are intercepted by the local `orted` via a pipe.  OS/pipe buffering may split single `fprintf()` messages into multiple strings, and therefore can (and do!) get munged together in `mpirun` stderr.  `opal_output()` and `opal_show_help()` guarantee to show their messages atomically (and without munging multiple messages together in `mpirun` stderr).
1. `opal_show_help()` de-duplicates error messages (i.e., only shows it once instead of N times, even if it is emitted multiple times from a single process, or even from multiple processes.

The de-duplication is surprisingly important.  It's the difference between seeing:

```
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
....etc.
```

and something like this:

```
Error returned by request event callback: 3.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
16 more processes ...
```

Reply:
I can switch to `opal_show_help()`. But having an issue when I switch to this- I am not seeing the help text output as it seems the process terminates sooner than the output stream can be flushed. I am a little puzzled by this as when I use `opal_output()` or `opal_output_verbose()`, I can see the output message. Internally they all seem to call `output()` (in opal/util/output.c) but `opal_show_help()` isn't working as expected. Any other internal mechanism to flush the output stream that I am missing?

Reply:
Ah; interesting point.  I wonder if we have a race condition here.

@rhc54 Can you shed any light on why `opal_output[_verbose]()` would tend to show up but `opal_show_help()` output would not when they are immediately followed by an `exit()`?  I'm guessing it's just a race condition; there's nothing explicit in `opal_output[_verbose]()` that ensures it is output before everything dies, right?

Reply:
opal_show_help goes thru PMIx log so we avoid making an OOB connection out of the app for aggregating the messages. Means you have to give PMIx time to transmit it to the daemon. 

opal_output just gets written thru a file descriptor and picked up by the daemon. fflush will force it out.

You should never call show_help or opal_output and then call exit! You are supposed to go thru finalize so the app gets an error. You never call exit from within a library (except when they call "MPI_Abort" and want us to do the exit for them).

Reply:
@rhc54 Thanks for the explanation of outputs.

This is actually a case where the application is totally hosed, and we're down deep in a PML or BTL.  I.e., the network stack is completely hosed and there is nothing rational left to do other than exit.  We even have a few calls to `exit(3)` in the usNIC BTL and elsewhere (they usually end up calling the upper-layer PML abort function, but sometimes not even that is available, and so they call `exit(3)` directly).

Reply:
Up to you - I only noted it because we have had complaints about doing that in the past. I know _we_ may believe there is nothing rational the user can do, but that assumes we really know that the app couldn't have done _anything_ useful (e.g., checkpoint partial results they have completed before exiting). Not sure the users out there necessarily agree with our assumption.

Reply:
@jsquyres , @rhc54 : Any further guidance on this?
FWIW, I tried using `opal_show_help()` followed by `return()` error value back to `opal_progress()` . The effect is that we still try to make progress even after this fatal error scenario. Example:

--------------------------------------------------------------------------
mtl_ofi.h:129: Call to OFI API fi_cq_read failed.
  Local host: arav-rhel
  Error: -258
  Error Description: Operation not permitted in current state
--------------------------------------------------------------------------
# OSU MPI-CUDA Bandwidth Test v5.3.2
# Send Buffer on HOST (H) and Receive Buffer on HOST (H)
# Size      Bandwidth (MB/s)
1                       1.17
2                       1.45
...
[arav-rhel:47807] 85721 more processes have sent help message help-mtl-ofi.txt / OFI call fail
2097152              1515.20
4194304              2417.16

I don't think we should be trying to make progress here. We should bail out (as Jeff mentioned- there is nothing rational left to do) . Given the absence of a better way to allow app to exit gracefully, it seems `opal_output()` followed by an `exit()` seems the best we have for this scenario. Thoughts?

Reply:
Sounds like our architecture leaves you without a good option here, so I guess we just have to exit. Be prepared for some user complaints, though I'm not familiar enough with this code area to offer an alternative.